src/coreclr/src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/ReadyToRunTableManager.cs

```diff
@@ -54,22 +77,97 @@ protected virtual void Graph_NewMarkedNode(DependencyNodeCore<NodeFactory> obj)
                 lock (_methodsGenerated)
                 {
                     Debug.Assert(!_sortedMethods);
-                    _methodsGenerated.Add(methodNode.Method);
+                    MethodDesc method = methodNode.Method;
+                    EcmaModule module = (EcmaModule)((EcmaMethod)method.GetTypicalMethodDefinition()).Module;
+                    if (!_methodsGenerated.TryGetValue(module, out var perModuleData))
+                    {
+                        perModuleData = new PerModuleMethodsGenerated(module);
+                        _methodsGenerated[module] = perModuleData;
+                    }
+                    if (method.HasInstantiation || method.OwningType.HasInstantiation)
+                    {
+                        perModuleData.GenericMethodsGenerated.Add(methodNode);
+                    }
+                    else
+                    {
+                        perModuleData.MethodsGenerated.Add(methodNode);
+                    }
                 }
             }
         }
-        public IEnumerable<MethodDesc> GetCompiledMethods()
+
+        public IEnumerable<IMethodNode> GetCompiledMethods(EcmaModule moduleToEnumerate, CompiledMethodCategory methodCategory)
         {
             lock (_methodsGenerated)
             {
                 if (!_sortedMethods)
                 {
                     TypeSystemComparer comparer = new TypeSystemComparer();
-                    _methodsGenerated.Sort((x, y) => comparer.Compare(x, y));
+                    Comparison<IMethodNode> sortHelper = (x, y) => comparer.Compare(x.Method, y.Method);
+
+                    List<PerModuleMethodsGenerated> perModuleDatas = new List<PerModuleMethodsGenerated>(_methodsGenerated.Values);
+                    perModuleDatas.Sort((x, y) => x.Module.CompareTo(y.Module));
```

Comment:
Completely unrelated, but do we have a way to sort collections using Schwartzian transform?

Reply:
Yes, there is a variant of Array.Sort which takes two arrays, a key array and a values array. It doesn't work for arbitrary collection types, but for consumers which are strongly motivated to use it, the fundamental sort algorithm is there to handle a transformed sort of that form.